File: services/backend-go/internal/readmodel/accounts.go
package readmodel

import (
	"errors"
	"math/big"
	"strings"
)

func validateAccounts(s Snapshot) error {
	if s.Accounts == nil {
		return nil
	}
	bad := errors.New("invalid account principal coverage")
	assets := map[string]string{}
	for _, c := range s.Configs {
		if c.Kind == "asset" {
			vault, _ := c.Values["userStockVault"].(string)
			assets[c.ID] = vault
		}
	}
	accounts := map[string]UserAccountReadModel{}
	totals := map[string]*big.Int{}
	end, err := Height(*s.Sync.BlockNumber)
	if err != nil {
		return bad
	}
	for _, a := range *s.Accounts {
		key := a.User + ":" + a.AssetUID
		if _, ok := accounts[key]; ok {
			return bad
		}
		at, err := Height(a.Source.BlockNumber)
		if a.Source.TransactionIndex > 1<<53-1 || a.Source.LogIndex > 1<<53-1 || err != nil || at > end || a.Source.ChainID != s.Sync.ChainID || a.User == "0x"+strings.Repeat("0", 40) || a.Vault == "0x"+strings.Repeat("0", 40) || assets[a.AssetUID] != a.Vault {
			return bad
		}
		deposited, e1 := raw(a.Deposited)
		allocated, e2 := raw(a.Allocated)
		free, e3 := raw(a.Free)
		if e1 != nil || e2 != nil || e3 != nil || deposited.String() != a.Deposited || allocated.String() != a.Allocated || free.String() != a.Free || deposited.Cmp(new(big.Int).Add(allocated, free)) != 0 {
			return bad
		}
		accounts[key] = a
		totals[key] = new(big.Int)
	}
	for _, p := range s.Positions {
		key := p.User + ":" + p.AssetUID
		a, ok := accounts[key]
		if !ok || a.Free != p.Free {
			return bad
		}
		n, err := raw(p.Allocated)
		if err != nil {
			return bad
		}
		totals[key].Add(totals[key], n)
	}
	for key, a := range accounts {
		if totals[key].String() != a.Allocated {
			return bad
		}
	}
	return nil
}

File: apps/web/src/v1/accounts.ts
import type {AccountPage, ConfigReadModel, SyncStatus, UserAccountReadModel} from './generated/read-api.ts';
const address=(v:unknown):v is `0x${string}`=>typeof v==='string'&&/^0x[0-9a-f]{40}$/.test(v)&&v!=='0x'+'0'.repeat(40);
const hash=(v:unknown):v is string=>typeof v==='string'&&/^0x[0-9a-f]{64}$/.test(v);
const uint=(v:unknown):v is string=>typeof v==='string'&&/^(0|[1-9][0-9]{0,77})$/.test(v)&&BigInt(v)<2n**256n;

// Commit the complete result only: a failed later page must not expose a partial wallet.
export async function loadWalletAccounts(options:{wallet:string;sync:SyncStatus;assets:readonly ConfigReadModel[];signal:AbortSignal;fetchPage:(cursor?:string)=>Promise<AccountPage>}):Promise<readonly UserAccountReadModel[]> {
 const {sync,signal}=options;
 const wallet=options.wallet.toLowerCase();
 const bad=()=>new Error('Account snapshot is unavailable or inconsistent');
 if(!address(wallet)||sync.status!=='synced'||sync.finality!=='finalized'||!uint(sync.blockNumber)||!hash(sync.blockHash)||sync.revision!==`${sync.blockNumber}:${sync.blockHash}`)throw bad();
 const vaults=new Map(options.assets.filter(a=>a.kind==='asset').map(a=>[a.id,a.values.userStockVault]));
 const rows:UserAccountReadModel[]=[];const cursors=new Set<string>();let cursor:string|undefined;let previous='';
 for(let page=0;page<100;page++) {
  if(signal.aborted)throw bad();
  const result=await options.fetchPage(cursor);
  if(signal.aborted)throw bad();
  if(!result||!result.sync||result.sync.chainId!==sync.chainId||result.sync.revision!==sync.revision||result.sync.blockNumber!==sync.blockNumber||result.sync.blockHash!==sync.blockHash||result.sync.status!=='synced'||result.sync.finality!=='finalized'||!Array.isArray(result.items)||result.items.length>100)throw bad();
  for(const a of result.items){
   if(!a||a.user!==wallet||!hash(a.assetUid)||a.assetUid<=previous||!address(a.vault)||vaults.get(a.assetUid)!==a.vault||![a.deposited,a.allocated,a.free].every(uint)||BigInt(a.deposited)!==BigInt(a.allocated)+BigInt(a.free))throw bad();
   const s=a.source;
   if(!s||s.chainId!==sync.chainId||!uint(s.blockNumber)||BigInt(s.blockNumber)>BigInt(sync.blockNumber)||!hash(s.blockHash)||(s.blockNumber===sync.blockNumber&&s.blockHash!==sync.blockHash)||!hash(s.transactionHash)||![s.transactionIndex,s.logIndex].every(n=>Number.isSafeInteger(n)&&n>=0))throw bad();
   previous=a.assetUid;rows.push(a);
  }
  if(result.nextCursor===null)return rows;
  if(typeof result.nextCursor!=='string'||result.nextCursor.length===0||result.nextCursor.length>1024||cursors.has(result.nextCursor)||result.items.length===0)throw bad();
  cursor=result.nextCursor;cursors.add(cursor);
 }
 throw bad();
}
